### ISSUE TITLE
Adding support for Smart Albums, where new media is added automatically based on conditions

### ISSUE THREAD

**Issue by cabe1214** [photoprism/photoprism#1431]
As a home user I'd like be able to create a Smart Album into which fotos will be added automatically depending on some condition(s).

i.e. I create a Smart Album "Baby" with the condition "Files which are located in Folder1". Every foto/video located/added to "Folder1" will be automatically added to the album "Baby" without having to insert the newly available files every time into the album.

Acceptance Criteria:
- [ ] "Create Smart Album" option MUST be listed when creating an album in the Albums section
- [ ] "Create Smart Album" option MAY be listed when creating an album based on a folder, adding the condition"File located in this folder".
- [ ] When creating smart albums it MUST be possible to specify one condition depending on which new fotos/videos will be added automatically to the album.
- [ ] It SHOULD be possible to specify several conditions of the same type linked with an OR operator (path A or B)
- [ ] It SHOULD be possible to specify several conditions of different type linked with an OR operator (path A or keyword K)
- [ ] It MAY be possible to link several conditions with AND or OR operators for complex conditional criteria.
- [ ] Smart Albums SHOULD be updated at least every 3 hours
- [ ] The user MAY be able to define a check interval (30 min, 1 hour, 3 hours, 1 day) for adding new fotos matching the conditions to the Smart Album.

**Comment by 10bn** [comment]
This would be great, I'd love to do location based smart albums!

**Comment by GlassedSilver** [comment]
For inspiration also checkk out Apple's various apps that have Smart Albums/Smart Playlists.

iTunes, Photos/iPhoto/Aperture, etc...

If there's one thing I have to continue to applaud Apple for it's smart folders being present in any app that has management of assets as its core feature.

To me, any kind of asset managing application is not complete without it. I think it's a CORE functionality.

**Comment by lastzero** [comment]
@GlassedSilver Apple has better funding. If we would have reached our funding goal, there might be a frontend for manually creating smart albums already. Sadly the work we do still doesn't pay our bills after 3.5 years... On the positive side, moments are smart albums so the functionality is already implemented in our backend. Just a nice UI for creating them manually is missing.

**Comment by GlassedSilver** [comment]
> 
> 
> @GlassedSilver Apple has better funding. If we would have reached our funding goal, there might be a frontend for manually creating smart albums already. Sadly the work we do still doesn't pay our bills after 3.5 years... On the positive side, moments are smart albums so the functionality is already implemented in our backend. Just a nice UI for creating them manually is missing.

I'm not trying to throw shade, just wanted to highlight the importance.

Sorry if I sounded rude in any way, that was not my intention at all! :o

**Comment by lastzero** [comment]
No worries, just trying to explain why seemingly simple or core features are not implemented yet. Even if we only needed one day for each issue, we would be busy for a year. Wish we had as many sponsors as feature requests.

**Comment by GlassedSilver** [comment]
> 
> 
> No worries, just trying to explain why seemingly simple or core features are not implemented yet. Even if we only needed one day for each issue, we would be busy for a year. Wish we had as many sponsors as feature requests.

I run a LOT of FOSS software, I definitely know what you mean.

I'm also scripting/coding small hobby projects in my spare time sometimes and my ideas always outpace my keyboard speed. :D 
You have my full sympathy. No software ever is done, all you can try is to close in on the distance between ideas and realization. :)

**Comment by snarfsmojo** [comment]
I love this idea!  an additional condition I'd like to see is people, specifically with the thought of sharing all of the pictures I have of an individual with them.

IE, based on the conditional value "contains person Steven Tyler" I can make a smart album named Steven Tyler, which I could then share with StevenTyler@aerosmith.com, :)

**Comment by joachimtingvold** [comment]
How will "Smart Albums" change the dynamic of `/unsorted` when implemented? Will photos "added" via "Smart Albums" show up or don't show up on `/unsorted`?

**Comment by lastzero** [comment]
Since Months are smart albums, there would be no unsorted pictures. Or if you manually add a smart album that doesn't filter (much). So probably makes no sense.

**Comment by GlassedSilver** [comment]
A smart album is a bit of a saved search, so... it's unsorted in my books, maybe it should be a user choice not sure.

**Comment by rybber78** [comment]
I'd love that too! It could be as simple as a saved search (perhaps with predefined sorting).

**Comment by GlassedSilver** [comment]
> I'd love that too! It could be as simple as a saved search (perhaps with predefined sorting).

plus maybe some metadata like setting a fixed album art or alternatively generating it from a selection of images dynamically.

**Comment by charles-997** [comment]
Another fantastic use-case for smart albums is "filtering" your best images vs simply all images.
I rate my images based on a 5 star system.  It would be remarkable if I were able to create smart albums based on the image's rating from the EXIF metadata.  I rate them in Lightroom and then when photoprism indexs, it picks up the metadata (I assume?).   A smart album would allow users to make use of this metadata to create a collection of "best of" images

**Comment by graciousgrey** [comment]
@charles-997 support for 5 star ratings is not yet implemented, but planned: https://github.com/photoprism/photoprism/issues/713  :)

**Comment by heitorPB** [comment]
Suggestion for the implementation: add a button (and endpoint as well?) to trigger updating the smart albums. That would be super handy when adding photos from external media. And might be helpful when implementing the last two items in the Acceptance Criteria list.

**Comment by GlassedSilver** [comment]
> Suggestion for the implementation: add a button (and endpoint as well?) to trigger updating the smart albums. That would be super handy when adding photos from external media. And might be helpful when implementing the last two items in the Acceptance Criteria list.

That should be done automatically anyhow, just like in iTunes for example. Nobody presses a button to update their personal Top 25 list.

Baffled by the suggestion that a smart album should be semi-automatic. o.O

**Comment by graciousgrey** [comment]
It's planned that smart albums should be updated automatically in the background, no button should be needed for this :)

**Comment by GlassedSilver** [comment]
> It's planned that smart albums should be updated automatically in the background, no button should be needed for this :)

Thanks for confirming, appreciated.

**Comment by heitorPB** [comment]
> That should be done automatically anyhow, just like in iTunes for example.

Would that work for network mounted filesystems? How would Photoprism detect file changes in an NFS share?

> Baffled by the suggestion that a smart album should be semi-automatic. o.O

I didn't say that =p I asked to have the functionality of updating the albums exposed to end users. Adding a button to click doesn't imply removing the automation. Those are complimentary.

In the case of remote storage of photos, the storage server can trigger a webhook when there are changes on disk. I can also configure my server to wait for the sync to complete before hitting the sync endpoint. Could also have a daily cron job downloading from Google Photos and if there are any new ones, trigger the smart-album update. The possibilities are endless.

Exposing the endpoint/button adds possibilities and flexibility. You don't _have to_ use if you don't want to.

> It's planned that smart albums should be updated automatically in the background,

I'm new to Photoprism, but does that mean updating only the smart-albums or re-syncing the library from disk? Would that read only from the database or re-scan the disk folders?

**Comment by charles-997** [comment]
> Would that work for network mounted filesystems? How would Photoprism detect file changes in an NFS share?

I mount the nfs share to the host system and then pass through to the PhotoPrism docker stack using a volume bind mount. Thus PhotoPrism is agnostic to local directory vs networked directory.

**Comment by GlassedSilver** [comment]
> > Baffled by the suggestion that a smart album should be semi-automatic. o.O
> 
> I didn't say that =p I asked to have the functionality of updating the albums exposed to end users. Adding a button to click doesn't imply removing the automation. Those are complimentary.

I don't mean to come across as a smart ass, but I didn't say manual, but semi-automatic. There's automation, semi-automation and manual handling. Semi-automation means that an automation is initiated based on user interaction. Think of paddle-shifting in racing cars. It's not manual because the clutch is handled by the car, and you only shift up and down, you don't directly select a gear with the car handling the specific selection.

But it's not automatic either, not fully at least, because if you didn't pull the paddle on your steering wheel you'd stick around in whatever gear you are.

> In the case of remote storage of photos, the storage server can trigger a webhook when there are changes on disk. I can also configure my server to wait for the sync to complete before hitting the sync endpoint. Could also have a daily cron job downloading from Google Photos and if there are any new ones, trigger the smart-album update. The possibilities are endless.
> 
> Exposing the endpoint/button adds possibilities and flexibility. You don't _have to_ use if you don't want to.

Even in a system that is supposed to work automatically usually workarounds are worked in that facilitate usage like you describe.
It's not untypical of folder-watching software to ALSO run a scheduled job to index or receive update commands.

A good example of that is my manga library software of choice: Kavita. :)

> > It's planned that smart albums should be updated automatically in the background,
> 
> I'm new to Photoprism, but does that mean updating only the smart-albums or re-syncing the library from disk? Would that read only from the database or re-scan the disk folders?

Re-scanning can possibly update any metadata that's new for photos and videos in your library, or, obviously, add new content that would then also possibly fall into criteria for a smart album. Re-syncing is a continuous process in software that has smart albums that's initiated whenever metadata for an item changes from within the software, so the item is checked for eligibility in smart folders by either removing it from them or adding it to them going by whatever the new metadata matches or doesn't match anymore for.

That's my understanding at least.

**Comment by kvalev** [comment]
A smart album will just be a saved query, so there will be little automation needed.

**Comment by GlassedSilver** [comment]
> A smart album will just be a saved query, so there will be little automation needed.

That plus caching, that's a technical difference. More than just a UX improvement.

**Comment by pacwoodson** [comment]
This is the most important feature I've been looking for in any photo library/digital asset management software. 
I even started to build my own app just for this, but didn't feel good building the nth photo library app out in the wild. 

No photo app have a feature like shareable smart albums with a prepared search query, this would be a killer one if Photoprism succeeds implementing this. 

What kind of work has to be done for this to work ? How is the backend ready for that ?

**Comment by lastzero** [comment]
@pacoccino I agree that having user-defined filters will be very helpful in many cases... The main things holding us back are (1) a missing UX concept, (2) having a secure implementation (as the filters could also be used to bypass restrictions if you are not an admin) and of course (3) the fact that we also have to work on other issues like 2FA and provide day-to-day support via email, chat and GitHub Discussions.

Adding an additional field to edit the applied filter is not difficult in itself. We just can't just release it like that for the reasons outlined above. So before we start developing it, it would be great to have a better understanding of the user interface and what else is needed, e.g. on the security side. Any help with this is welcome!

As a quick hack, I think you could e.g. create a "moment" album in the database, set a manual filter via SQL or a database management UI and then share that...

Edit: If you are familiar with JS/CSS, an autocomplete functionality for filters in the existing search bar could be a start, as many users struggle to remember filter names and some don't even know they exist. We could then move from there?

For an overview of the filters you can use:
- https://docs.photoprism.app/user-guide/search/filters/#filter-reference

**Comment by pseudocoder** [comment]
This would be a great addition. Key thing for me would be the ability to share a smart album - I can always run a search myself, and even bookmark it as an alternative to a smart album, but I really want to be able to share a search via a link which this would provide.

**Comment by rheckly** [comment]
Is there a way to create an album based on a filter / search pattern?
I don't want to select manually the pics from the search result - but have the filters as defined for an album content.
changing the filter would change the pictures in that album.
Similar to search features in plex music app.

**Comment by GlassedSilver** [comment]
> Is there a way to create an album based on a filter / search pattern? I don't want to select manually the pics from the search result - but have the filters as defined for an album content. changing the filter would change the pictures in that album. Similar to search features in plex music app.

If there was we wouldn't have this tracked issue. 😉

**Comment by kvalev** [comment]
> Is there a way to create an album based on a filter / search pattern?

Yeah, in theory only the UI is missing, so you create an empty album and change the `filter` field in the database to a valid search expression, e.g.

`favorite:true country:sc year:2018`

**Comment by rheckly** [comment]
Thank you for this hint
I'm now connected to the database and updated the field albums.album_filter with:
update photoprism.albums set album_filter = _binary 'people:Roger'  where album_title = 'Roger' and id = 791;

by using this select, I see the value in the field:
SELECT id, album_title, CAST(album_filter AS CHAR) AS album_filter
FROM photoprism.albums
WHERE  id = 791;

But wondering, that the album shows not only pics where Roger is on or face detected result is Roger, but also a lot other pics.
Do I have to add something in the filter?

When I search in Originals, the filter works.

**Comment by rheckly** [comment]
> Thank you for this hint I'm now connected to the database and updated the field albums.album_filter with: update photoprism.albums set album_filter = _binary 'people:Roger' where album_title = 'Roger' and id = 791;
> 
> by using this select, I see the value in the field: SELECT id, album_title, CAST(album_filter AS CHAR) AS album_filter FROM photoprism.albums WHERE id = 791;
> 
> But wondering, that the album shows not only pics where Roger is on or face detected result is Roger, but also a lot other pics. Do I have to add something in the filter?
> 
> When I search in Originals, the filter works.

I tried with filter person:Roger but also with people:Roger - but both without success.

Anybody has an idea, why this doesn't work? I also enabled Beta Features. There are no Pictures in table photos_albums for this Album - but it shows a lot pictures.

**Comment by kvalev** [comment]
> > Thank you for this hint I'm now connected to the database and updated the field albums.album_filter with: update photoprism.albums set album_filter = _binary 'people:Roger' where album_title = 'Roger' and id = 791;
> > by using this select, I see the value in the field: SELECT id, album_title, CAST(album_filter AS CHAR) AS album_filter FROM photoprism.albums WHERE id = 791;
> > But wondering, that the album shows not only pics where Roger is on or face detected result is Roger, but also a lot other pics. Do I have to add something in the filter?
> > When I search in Originals, the filter works.
> 
> I tried with filter person:Roger but also with people:Roger - but both without success.
> 
> Anybody has an idea, why this doesn't work? I also enabled Beta Features. There are no Pictures in table photos_albums for this Album - but it shows a lot pictures.

This is the query I am using, but I have my own photoprism fork, so I dont know if its supported upstream.

> subjects:"My Name & Another Name"

**Comment by rheckly** [comment]
> > > Thank you for this hint I'm now connected to the database and updated the field albums.album_filter with: update photoprism.albums set album_filter = _binary 'people:Roger' where album_title = 'Roger' and id = 791;
> > > by using this select, I see the value in the field: SELECT id, album_title, CAST(album_filter AS CHAR) AS album_filter FROM photoprism.albums WHERE id = 791;
> > > But wondering, that the album shows not only pics where Roger is on or face detected result is Roger, but also a lot other pics. Do I have to add something in the filter?
> > > When I search in Originals, the filter works.
> > 
> > 
> > I tried with filter person:Roger but also with people:Roger - but both without success.
> > Anybody has an idea, why this doesn't work? I also enabled Beta Features. There are no Pictures in table photos_albums for this Album - but it shows a lot pictures.
> 
> This is the query I am using, but I have my own photoprism fork, so I dont know if its supported upstream.
> 
> > subjects:"My Name & Another Name"

Thank you very much kvalev! This helped! In the photoprism help https://docs.photoprism.app/user-guide/search/filters/ i just looked for the filters People /Person - now i see there are aliases for subject / subjects - with the same definition as people / person.
So smart album filters must be done with subject or subjects.
